### PR TITLE
Ghost - Refactor away from player to HUD/Gamerule, PVS for smoother visual

### LINKED ISSUE #475: Refactor: `CNEOHud_GhostMarker` should internally work out ghost states
Currently some setters set by a `C_NEO_player` to pass on ghosting team, client's states, etc.. but it can be done within `CNEOHud_GhostMarker` instead. Mostly less possible state issues.

### LINKED ISSUE #536: [BUG] Ghost holder's icon is ahead of player's movement
### Build Info

20240801-9d04458

### Description

Ghost holder's HUD icon moves ahead of the player's movements. So leaning left and moving left results in an icon slightly ahead of the player's actual position. Alternating left and right lean while moving makes icon sway a lot more than OG.

### To Reproduce

Pickup ghost and lean while walking in front of viewer.

### Expected behavior

Ghost marker should be accurate to the player's position.

### Actual behavior

Ghost marker is ahead of the player when leaning and walking by several units

### Operating System

- [X] Windows
- [ ] Linux

### Version/Distro

Windows 11

### Machine's CPU

AMD Ryzen 9 5900X

### Machine's GPU

NVIDIA GeForce RTX 3080 Ti

### GPU's driver

_No response_

### Build's compiler

_No response_

### Additional context and Screenshots

[ghost_indicator.webm](https://github.com/user-attachments/assets/3fec9e38-b3af-4708-a02b-93086f62f74b)

### PR DESCRIPTION
**Pull request by nullsystem**
## Description

* **Reviewers: Basically if the ghost HUD looks smooth rather than jittery, then the PVS is active for the ghost**
* No longer handles states from player to HUD
* HUD handles some states itself, some given to gamerules to handle it instead especially for networked
* PVS over networked for smoother ghost visuals
    * Ghost is now always in PVS
* Keep pointer to ghost entity:
    * Server/gamerule: From first spawn - to not need to refetch on each think
    * Client/HUD: From first player ghoster - always in PVS, smoother visuals
    * Both removes pointer once the round ends

## Toolchain

- Linux GCC Distro Native Arch/GCC 14

## Linked Issues

* fixes #475
* fixes #536

### CODE REVIEW

**Comment by AdamTadeusz** [comment]
Ghost is buttery smooth when in pvs, looks great. Would it be worth using the same text fadeout distance value for the ghost marker as for the capture zone marker?

**Comment by nullsystem** [comment]
@AdamTadeusz I've now put that in as its own convar, but same default as the cap zone's default.

File: mp/src/game/shared/neo/neo_gamerules.cpp [hunk 1]
```diff
@@ -536,20 +542,73 @@ void CNEORules::Think(void)
...
+				Assert(nextGhosterTeam == TEAM_JINRAI || nextGhosterTeam == TEAM_NSF);
+				break;
+			}
 		}
 	}
 	m_iGhosterTeam = nextGhosterTeam;
 	m_iGhosterPlayer = nextGhosterPlayer;
 
+	static int ghostEdict = -1;
+	if (!ghost)
```
[the remark this answers is not in the document]
Comment (AdamTadeusz) [reply mp/src/game/shared/neo/neo_gamerules.cpp:564]: Takes about 50 microseconds every game tick to find the ghost with 15 players. Can reduce it down to 0-1 microseconds by instead keeping a pointer to the ghost which gets updated when the ghost is reset on map change etc,  but struggling to fix an issue with that method so gonna approve this for now

**Review by AdamTadeusz — DISMISSED** [review]

File: mp/src/game/shared/neo/neo_gamerules.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (Rainyan) [reply mp/src/game/shared/neo/neo_gamerules.cpp:564]: gEntList.FindEntityByClassname is an option, although I haven't measured it

[the remark this answers is not in the document]
Comment (nullsystem) [reply mp/src/game/shared/neo/neo_gamerules.cpp:564]: I don't think it's every single Think it'll go through the whole of this due to the static int? Nevertheless yes, the section can be better.

**Comment by nullsystem** [comment]
I'll see if I can refactor this further, I'm putting this in draft for now.

File: mp/src/game/shared/neo/neo_gamerules.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (AdamTadeusz) [reply mp/src/game/shared/neo/neo_gamerules.cpp:564]: I printed the the difference between the start and end to console and it was around 50 microseconds every single think

[the remark this answers is not in the document]
Comment (nullsystem) [reply mp/src/game/shared/neo/neo_gamerules.cpp]: I've refactored the spawning and think section relating to the ghost. It now keeps a pointer to it, this is also applied to the client-side HUD also for further PVS-usage improvement so smooth visual is applied from the first ghost pickup.

**Comment by nullsystem** [comment]
With the further refactor in, there's a bit of an overlap with #514 as opted to don't visualize the ghost on post-round. But PVS is further improved now with always being used and used since first ghoster.

**Comment by nullsystem** [comment]
I'm allowing this for review, but I'll be putting PR #511 for merging in before this.

**Comment by AdamTadeusz** [comment]
conflicts

**Comment by nullsystem** [comment]
@AdamTadeusz fixed

**Comment by VantoNortim** [comment]
Would it not be easier to set the ghost to always stay networked? If the position is not correct when picked up we could network the player who is carrying the ghost too.

**Comment by nullsystem** [comment]
> Would it not be easier to set the ghost to always stay networked? If the position is not correct when picked up we could network the player who is carrying the ghost too.

@VantoNortim It is changed to stay always networked including the player ghosting. The server set position is a "backup" positioning although may not be used.

**Review by AdamTadeusz — APPROVED** [review]

**Review by AdamTadeusz — APPROVED** [review]

**Review by brysondev — APPROVED** [review]
LGTM

**Pull request merged**